src/Morphir/Spark/AST.elm

```diff
@@ -220,6 +274,139 @@ namedExpressionsFromFields ir fields =
         |> ResultList.keepFirstError
 
 
+objectExpressionFromAggregationCall : IR -> AggregationCall -> Result Error ObjectExpression
+objectExpressionFromAggregationCall ir aggregationCall =
+    case aggregationCall of
+        AggregationCall groupKey _ aggValues sourceRelation ->
+            -- not using returned group key yet
+            Result.map3
+                Aggregate
+                (groupKey |> Name.toTitleCase |> Ok)
+                (aggValues
+                    |> List.map
+                        (\aggValue ->
+                            case aggValue of
+                                AggregateValue fieldName _ aggName (Just aggArg) ->
+                                    mapSDKFunctions ir [ aggArg ] aggName
+                                        |> Result.andThen (\expr -> Tuple.pair fieldName expr |> Ok)
+
+                                AggregateValue fieldName _ aggName Nothing ->
+                                    mapSDKFunctions ir [] aggName
+                                        |> Result.andThen (\expr -> Tuple.pair fieldName expr |> Ok)
+                        )
+                    |> ResultList.keepFirstError
+                )
+                (objectExpressionFromValue ir sourceRelation)
+
+
+{-| Creates an Aggregate ObjectExpression from source values
+-}
+constructAggregationCall : TypedValue -> TypedValue -> TypedValue -> Result Error AggregationCall
+constructAggregationCall aggregateBody groupKey sourceRelation =
+    let
+        getKeyField : TypedValue -> Result Error (Maybe Name)
+        getKeyField body =
+            case body of
+                -- extract any field name that "key" is used for to check if the groupBy column is renamed
+                Value.Lambda _ (Value.AsPattern _ _ keyName) (Value.Lambda _ (Value.AsPattern _ _ inputsName) (Value.Record _ fields)) ->
+                    let
+                        keyFields =
+                            fields
+                                |> Dict.toList
+                                |> List.filter
+                                    (\( name, value ) ->
+                                        case value of
+                                            Value.Variable _ varName ->
+                                                varName == keyName
+
+                                            _ ->
+                                                False
+                                    )
+                    in
+                    case keyFields of
+                        -- expect only one field that uses "key"
+                        ( name, _ ) :: [] ->
+                            Just name |> Ok
+
+                        [] ->
+                            Nothing |> Ok
+
+                        _ :: _ :: [] ->
```

Comment:
A simpler way of writing this would be: `[ _, _ ]`. Is there a reason why having exactly 2 key field usages should be reported differently than having more than 2 key field usages? Maybe it would be easier to just remove this branch and let the last case catch all of these?

Reply:
I think I misunderstood this as meaning _at least two key fields_, on the assumption that `::` also concatenated lists together instead of only appending values to a list.

Reply:
You didn't completely misunderstand it, you can use this to do "at least two" but you need to change it slightly:

```elm
_ :: _ :: _ ->
```

The difference is that the last pattern which always refers to the tail (in other words, all the remaining elements in the list) is a wildcard here instead of an empty list. 

Also, slight correction: `::` prepends instead of appending and behaves the same in  pattern match. It's easier to think of it as `head :: tail` where "head" refers to the first element and "tail" refers to all the remaining elements. And it's composable so you can do: `first :: (second :: (third :: rest))` and thanks to `::` being [right-associative](https://github.com/elm/core/blob/84f38891468e8e153fc85a9b63bdafd81b24664e/src/List.elm#L41) you don't have to add parentheses because that's how it's interpreted when you do `first :: second :: third :: rest` anyway. Pretty clever design.

Reply:
Thanks! I was slightly concerned that there was no "list of any size" constructor I could use instead of `_` at the end to enforce its type, but it appears `::` is exclusively for type `a -> List a -> List a`, so that works out fine :)

Reply:
The language design of Elm makes it unnecessary to enforce types in a pattern. Simply put, everything you match on has to have one specific concrete type in order for the code to compile. So you can't even match on anything else other than a list here. That's why there are no type checking operators in the language at all. Constructors make it seem like you are doing a type check but constructors are values of the same type so it's not a type check. There is no inheritance so there is no need to check for a specific type, it's always specific :)